### PR TITLE
(CDAP-7420) Program lifecycle impersonation

### PR DESCRIPTION
**Pull request by chtyim**
Refactor the ProgramRuntimeService and wrap the TwillController to minimize places that need to be aware of impersonation

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java [hunk 1]
```diff
@@ -250,14 +278,18 @@ public boolean apply(RunRecordMeta record) {
...
+      ProgramController programController = createController(programDescriptor, twillController, runId);
+      return programController == null ? null : new SimpleRuntimeInfo(programController,
+                                                                      programId.toEntityId(),
+                                                                      twillController.getRunId());
     } catch (Exception e) {
       return null;
     }
   }
 
+  @Nullable
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java:292]: does it actually ever return null?

File: cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillController.java [hunk 2]
```diff
@@ -333,9 +326,39 @@ private boolean hasRun(YarnApplicationState state) {
...
+    }
+    synchronized (this) {
+      if (resourcesClient != null) {
+        return resourcesClient;
+      }
+
+      YarnApplicationReport report = processController.getReport();
+      String host = report.getHost();
+      int port = report.getRpcPort();
+      if (host == null || host.equals("N/A") || port == -1) {
```
Comment (anwar6953) [inline cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillController.java:346]: I think we should LOG#warn in this case.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedProgramRuntimeService.java:292]: According to the logic, it's possible. And the caller also checks for it, hence I add the annotation to be clear.

File: cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillController.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/org/apache/twill/yarn/YarnTwillController.java:346]: fixed

**Comment by anwar6953** [comment]
http://builds.cask.co/browse/CDAP-RUT227-1

**Comment by anwar6953** [comment]
LGTM; tested on cluster. Squashing commits.

**Comment by anwar6953** [comment]
Build green; merging

**Pull request merged**